### ISSUE TITLE
Bug: [no-unnecessary-type-parameters] False positive when type parameter is used once in member of class

### ISSUE THREAD

**Issue by reduckted** [typescript-eslint/typescript-eslint#9961]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&fileType=.ts&code=JYOwLgpgTgZghgYwgAgGpwDYFcIElzTxIA8AKgHzIDeAsAFDLIBumOAXMqQNyP0C%2B9eggxwAzqLSsIAYRHiylWg2QAHLACMMwBMykdu-QXWFiJ6bDLmiA4hDCQoC6vUZrN25AHM7uiwAoASn1nZUYoOywoEGQQLAwMAEIuF2QBOjShK0kLWVMAWTsACwB7ABMnJVcNLR1vMHMcQODKxmRwsEjo2PiklLS0oA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWsRFCAtonyJoqDJADuIppHBgAviF1A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
interface ValueInterface<T> { // Correctly not reported.
  value: T;  
}

class ValueClass<T> { // False positive.
  public value: T;
}

class ValueClassGetter<T> { // False positive.
  public get value(): T {
    return null!;
  }
}

class ValueClassMethod<T> { // False positive.
  public getValue(): T {
    return null!;
  }
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  "rules": {
    "@typescript-eslint/no-unnecessary-type-parameters": "warn"
  }
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

No warnings are reported.

### Actual Result

A warning is reported for the type parameter on each class.

```
Type parameter T is used only once in the class signature. 5:18 - 5:19
Type parameter T is used only once in the class signature. 9:24 - 9:25
Type parameter T is used only once in the class signature. 15:24 - 15:25
```

**Comment by abrahamguo** [comment]
Can you explain the difference between the class and the interface?

Do we allow single-use type parameters on any class, or what should be the corrected logic here?

**Comment by auvred** [comment]
If we allow to write

```ts
class ValueClass<T> {
  value: T
}
```

then we should allow to write

```ts
function makeValueClass<T>(): {
  value: T
} {}
```

Because it basically means the same thing. 


This code makes sense only if we assume that class user will manually specify `T` at class instance creation time:

```ts
const instance = new ValueClass<string>()
instance.value
//        ^? string
```

The same applies to `makeValueClass` mentioned above: the signature of this function makes sense only under this assumption. But the rule intentionally forbids `makeValueClass`-like functions. So I think the current behavior of the rule is correct.

Any thoughts?

**Comment by auvred** [comment]
Oh, there is already discussion about this particular case: https://github.com/typescript-eslint/typescript-eslint/pull/9536#discussion_r1713663117. And it doesn't look like an agreement has been reached yet.

**Comment by bradzacher** [comment]
Note that there is a massive difference between your two examples.

The function form could reasonably be converted to `unknown` without any loss of safety or change in functionality both inside and outside of the function. If the generic has a constraint, the same applies just using the constraint instead of `unknown`.

The same is not true of a class with a generic property. The generic persists into the class instant and also the property (if public) is an outwardly observable thing, unlike a function type parameter which is only inwardly observable.

**Comment by auvred** [comment]
> The function form could reasonably be converted to `unknown` without any loss of safety or change in functionality both inside and outside of the function.

IMO we can't remove type parameter from some factory functions without losing functionality. Function can be used to construct some object with constraints, just like a class constructor is used to construct an instance of a class with type constraints.

Consider the following example, I think the class and function here should behave the same way:

```ts
declare class Logger<T extends 'info' | 'debug'> {
  log(level: T): void
}

declare function createLogger<T extends 'info' | 'debug'>(): {
  log(level: T): void
}

const infoLogger_class = new Logger<'info'>()
infoLogger_class.log('info')
infoLogger_class.log('debug')
//                   ~~~~~~~
//   2345: Argument of type '"debug"' is not assignable to parameter of type '"info"'.


const infoLogger_function = createLogger<'info'>()
infoLogger_function.log('info')
infoLogger_function.log('debug')
//                      ~~~~~~~
//   2345: Argument of type '"debug"' is not assignable to parameter of type '"info"'.
```

Both `Logger` and `createLogger` create an object with a `log` method that accepts only the log level specified when the object was created.

**Comment by bradzacher** [comment]
The latter is a clear violation of the rule - you're using a generic in a return location only. In this case it's "kind of safe" cos the usecase is so very constrained - but it's not really any better than a function like `foo<T>(): T`.

The former is the same case as the other one if you squint. A function that creates an object is kinda the same as a class (which is a special function that creates a special object via a special syntax). But we classify the class as different because the intent is clearly different and the safety is much different - the scope for abuse and unsafety is much smaller.

**Comment by JoshuaKGoldberg** [comment]
@kirkwaiblinger and I had a long discussion on this. In summary, we'd like to propose adding off-by-default opt-in options for parameters of fields/methods of returned _classes_ and _functions_. Vague straw man proposal starting names:

* **`allowClassMemberParameters`**: for cases like [`class Logger<T> { log(level: T): void; }`](https://typescript-eslint.io/play/#ts=5.6.2&fileType=.ts&code=CYUwxgNghgTiAEkoGdnwDIHsDm2QwB4AVeEADwBcQA7YNAclACMBXbe%2BAH3noEtqAZpnoA%2BeAG8AUPHgQcACgggAbiAgAueEQCUm5Zl7BJAX0mSwmasgrxmbLLnzwAvPGogA7hhx5CjEKzsIvLaANxmdtgOvgB0ctjyAET8QolhEQH2PvhxComRaeHmltbwKZjRTq7uXpV%2B5aIhReV1uQnJgpiFZi3ZMG1JBelAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWsRFCAtonyJoqDJADuIppHBgAviF1A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)
* **`allowFunctionReturnParameters`**: for cases like [`function createLoggerInline<T>(): { log(level: T): void; }`](https://typescript-eslint.io/play/#ts=5.6.2&fileType=.ts&code=FAEwpgxgNghgTmABAMwK4DsIBcCWB7dRCBGLMAGTwHMqw4BJdKHdMAHgBVEwAPM9EAGdEAchbI8IxAB9R4AEaoqIgHwAKAJQAuRAG9giRFGpqoYAG5goOjtsTm8OEMAC%2BAbmDAIBQVkTi8Sho6RABeIhIyINoGJhZ2MXQJVU0PYADougA6Yyo1ACIA-I00jOoYnJN8hSVitO90X0Qaqky4MIiwUgpyukZmVjYRFpSSzxa2yrzCpLw68bBFVt64KYKWuqA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWsRFCAtonyJoqDJADuIppHBgAviF1A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

There's a _lot_ of complexity and potential for edge cases. So we're hoping that those two options can get users _most of the way_ there as heuristics that balance being roughly what folks are looking for & also being somewhat straightforward to explain.

cc @bradzacher @danvk @Josh-Cena 

---

Both _functions_ (i.e. instantiation expressions) and _classes_ share the following quirk. Namely, that a type parameter used once and then as the type of a method or property _can_ in the return / method return help make created objects more specific. 

For both of them, there are valid use cases - as mentioned in this issue and https://github.com/typescript-eslint/typescript-eslint/pull/9536/files#r1705850744. We're seeing a class as essentially being a function that returns an object and has to be called with `new`. The conceptual parallels are:
* Class constructor parameters for a class are parallel to function parameters
* Class fields are parallel to functions' returned object fields

Note that those options are specific to _parameters_ of the methods and function fields. We don't want to allow:

```ts
declare class Logger<T extends 'debug' | 'info'> {
  log(): T;
}
```

...as it doesn't map any inputs to the returned `T`. It's essentially the same flaw as `function parseJSON(input: string): T` with extra steps.

We also looked at the currently-not-reported-upon case of having _two_ methods that return a `T`:

```ts
declare class Logger<T extends 'debug' | 'info'> {
  log1(): T;
  log2(): T;
}
```

We think that that is a separate issue. That's about whether `T` is being detected as only ever used in a return. This issue is about `T` being used to narrow what consumers can pass into the created instance/object's fields.

Another separate issue is mutable fields, such as:

```ts
declare class Box<T> {
  value: T | undefined;
}

declare function createBox<T>(): { value: T };
// (also, for fun...)
declare function createBoxFactory<T>(): () => { value: T };
declare function createBoxFactoryFactory<T>(): () => () => { value: T };
```

In those cases, we think that's a separate issue. The `value` property is essentially two things: a getter and a setter. So the `T` does make a difference.

We _also_ looked at two more cases that kind of combine everything:

```ts
declare class Logger1<T extends 'debug' | 'info'> {
  log1(): () => T;
}

declare class Logger2<T extends 'debug' | 'info'> {
  log2(): (input: T) => void;
}
```

* `Logger1` is that separate bug issue: it doesn't restrict the way you use the returned object.
* `Logger2` _is_ what this issue is talking about. It restricts the way you use the returned object - just with an extra step.

In other words, if the type parameter is used in _any_ function parameter position somewhere nested within the type, then the proposed option would allow it.

Shoutout @kirkwaiblinger for ideating all those glorious edge cases.

Co-authored-by: @kirkwaiblinger

**Comment by danvk** [comment]
> Note that those options are specific to _parameters_ of the methods and function fields. We don't want to allow:
> 
> ```ts
> declare class Logger<T extends 'debug' | 'info'> {
>   log(): T;
> }
> ```
> 
> ...as it doesn't map any inputs to the returned `T`. It's essentially the same flaw as `function parseJSON(input: string): T` with extra steps.

What would help me is to see the suggested rewrite for this and the [class Logger<T> { log(level: T): void; }](https://typescript-eslint.io/play/#ts=5.6.2&fileType=.ts&code=CYUwxgNghgTiAEkoGdnwDIHsDm2QwB4AVeEADwBcQA7YNAclACMBXbe%2BAH3noEtqAZpnoA%2BeAG8AUPHgQcACgggAbiAgAueEQCUm5Zl7BJAX0mSwmasgrxmbLLnzwAvPGogA7hhx5CjEKzsIvLaANxmdtgOvgB0ctjyAET8QolhEQH2PvhxComRaeHmltbwKZjRTq7uXpV%2B5aIhReV1uQnJgpiFZi3ZMG1JBelAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWsRFCAtonyJoqDJADuIppHBgAviF1A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false) example. They seem different than `JSON.parse` in that you bind the type parameter once and then call the method many times. In that sense, they are more like instantiation expressions.

**Comment by JoshuaKGoldberg** [comment]
```ts
declare class Logger {
  log(): 'debug' | 'info'
}
```

```ts
declare class Logger {
  log(level: 'debug' | 'info'): void;
}
```

**Comment by bradzacher** [comment]
Classes are a funny thing with type parameters.
They don't necessarily follow the same "golden rule" as functions.
For example it's completely valid to define a class that's parameterised just once eg:
```ts
declare class Logger<T> {
  log(level: T): void;
}
```
Because whilst the parameter is used exactly once - it does change the behaviour and constraints of the class. EG:
```ts
const debug = new Logger<'debug'>();
debug.log('debug'); // ✅ no error
debug.log('info'); // ✅ errors

// which is different to a similar looking function
declare function logger<T>(level: T): void;
logger('info'); // no error
logger('debug'); // no error
// why would you ever do this?
logger<'debug'>('info');
```

However it's also not valid to write something like this, because it's physically impossible to safely implement this class
```ts
class UnsafeLogger<T> {
  log(): T;
}

// just as impossible to safely implement as
declare function unsafeLogger<T>(): T;
```

---

I think that a major thing the rule needs to separate is the variance of the usage of the type parameter.

```ts
class Covariant<T> {
  method(): T;
}
function covariant<T>(): T;

class Contravariant<T> {
  method(value: T): void;
}
function contravariant<T>(value: T): void;
```

If a type parameter is only used in a covariant position then it is going to be an unsafe type parameter in that it is impossible to implement safely (i.e. there's going to be some air of "trust me bro this is correct" to the code).
This applies to both classes and functions alike.

In a class if a type parameter is only used in a contravariant position then it always going to be a safe and necessary type parameter, regardless of how many times it is used. The rule should **_ALMOST_** never report on this case.

Note that I say almost because there is one case where the rule should report - if the type parameter is used just once AND that usage is in the constructor. Why is this usage invalid but the usage in a method isn't invalid? Because classes are syntactic sugar and you can think of a class as a function that returns an object. So when you adjust the definition appropriately and it becomes immediately obvious
```ts
class UnnecessaryParameter<T> {
  constructor(arg: T);
}
// becomes
function UnnecessaryParameter<T>(arg: T): {};
```
Similarly if you look at the contravariant case:
```ts
class Contravariant<T> {
  method(value: T): void;
}
// becomes
function Contravaraint<T>(): {
  method(value: T): void;
};
```

In a function if a type parameter is used once in a contravariant position then that type parameter is necessary if and only if the usage is not in the function's arguments -- i.e. contravariant usage in the return type is necessary.

I don't believe there should be the options you described @JoshuaKGoldberg -- because IMO they should just be the default behaviour for the reasons I laid out above.

**Comment by kirkwaiblinger** [comment]
If someone can implement the type parameter variance analysis, I'm in favor of it happening. I think it seems tedious/difficult to do so.

It does perhaps let us beneficially recast the discussion around special-casing "single-use" types as "types to consider to be covariant despite being mutable"
